Define JavaScript runtimes semantic conventions (#2290)
open-telemetry/opentelemetry-specification: specification/resource/semantic_conventions/process.md

### JavaScript runtimes

> JavaScript instrumentation should fill in the values by copying from built-in runtime constants.

Reviewer: Please update the CHANGELOG.

- `process.runtime.name`:
  - When the runtime is Node.js, fill in the constant value `nodejs`.
  - When the runtime is Web Browser, fill in the constant value `browser`.
- `process.runtime.version`:
  - When the runtime is Node.js, fill in the value of `process.versions.node`.
  - When the runtime is Web Browser, fill in the value of `navigator.userAgent`.
- `process.runtime.description`:

>   - When the runtime is Node.js, fill in the values of `"Node.js"`, `process.version` in that order, separated by spaces.

Reviewer: ```suggestion
  - When the runtime is Node.js, fill in the values of `"Node.js v{process.runtime.version}"` where `{process.runtime.version}` is replaced accordingly.
```
Otherwise the `v` you use in your example is missing.

Reviewer 2: Please add an example to make it clear what the concatenated value looks like.

Author: In Node.js, `process.version` outputs the version strign with the leading `v`: `"v14.15.4"`, whilst the versions in `process.versions.node` are not: `"14.15.4"`. This is intentional:

```bash
$ node --print "process.version"
v14.15.4
$ node --print "process.versions.node"
14.15.4
```

Reviewer 3: Maybe we can update the wording to make that more clear to people who aren't js devs, but i wouldn't expect people unfamiliar with js to need this spec so it might not be a big deal?

>   - When the runtime is Node.js, fill in the values of `"Node.js"`, `process.version` in that order, separated by spaces.

Reviewer 3: ```suggestion
  - When the runtime is Node.js, fill in the values of `"Node.js"`, `process.version` in that order, separated by a space.
```

>   - When the runtime is Node.js, fill in the values of `"Node.js"`, `process.version` in that order, separated by spaces.

Reviewer 4: Does this resource attribute add any new information? The runtime name and version are already reported on the other attributes.
Since it's not required, maybe it's best to just not fill it in node?

Author: No, there is no information like binary build time available in Node.js (also no additional information for web browsers). Omitting sounds good to me.

### JavaScript runtimes

JavaScript instrumentation should fill in the values by copying from built-in runtime constants.

- `process.runtime.name`:
  - When the runtime is Node.js, fill in the constant value `nodejs`.
  - When the runtime is Web Browser, fill in the constant value `browser`.
- `process.runtime.version`:
  - When the runtime is Node.js, fill in the value of `process.versions.node`.

>   - When the runtime is Web Browser, fill in the value of `navigator.userAgent`.

Reviewer 5: I heard userAgent is on a path to getting deprecated nowadays? https://blog.chromium.org/2021/09/user-agent-reduction-origin-trial-and-dates.html
Also, an uparsed userAgent might be better suited as process.runtime.description.

Author: The document here for userAgent is not essentially changed: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/process.md#javascript-runtimes. This PR writes down the details of other JavaScript runtime attributes.

It is true that chrome has published the deprecation path of `navigator.userAgent`. However, the fact is, the new experimental API is not widely available on all web platforms: https://caniuse.com/mdn-api_navigator_useragentdata. So I believe we are not in a hurry to update our conventions here.

Reviewer 3: @Oberon00 IMO the response is acceptable to resolve this conversation
